architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,693 @@
+	BicepDeploymentPhraseReady BicepDeploymentPhrase = "Ready"
+
+	// BicepDeploymentPhraseFailed indicates that the Bicep Deployment has failed.
+	BicepDeploymentPhraseFailed BicepDeploymentPhrase = "Failed"
+
+	// BicepDeploymentPhraseDeleting indicates that the Bicep Deployment is being deleted.
+	BicepDeploymentPhraseDeleting BicepDeploymentPhrase = "Deleting"
+
+	// BicepDeploymentPhraseDeleted indicates that the Bicep Deployment has been deleted.
+	BicepDeploymentPhraseDeleted BicepDeploymentPhrase = "Deleted"
+)
+```
+
+#### BicepDeploymentResource Custom Resource Definition
+
+The `BicepDeploymentResource` CRD is another CRD that will be responsible for tracking the state of individual resources. The CRD will have the following fields:
+
+```go
+// BicepDeploymentResourceSpec defines the desired state of a BicepDeployment
+type BicepDeploymentResourceSpec struct {
+  // ResourceID is the resource ID of the resource to deploy.
+  ResourceID string `json:"resourceID"`
+}
+
+type BicepDeploymentResourceStatus struct {
+  // ObservedGeneration is the most recent generation observed for this BicepDeploymentResource.
+  ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+  // Operation tracks the status of an in-progress provisioning operation.
+  Operation *ResourceOperation `json:"operation,omitempty"`
+
+  // Phrase indicates the current status of the Bicep Deployment Resource.
+  Phrase BicepDeploymentResourcePhrase `json:"phrase,omitempty"`
+
+  // Description is a human-readable description of the status of the Bicep Deployment Resource.
+  Description string `json:"description,omitempty"`
+}
+```
+
+#### `rad bicep generate-kubernetes` Command
+
+The `rad bicep generate-kubernetes` command will be a new command that will generate a Kubernetes custom resource when provided a Bicep template and its parameters. The command will take the following arguments:
+
+- `-p, --parameters`: The parameters for the Bicep template. Can be specified in the same way as [`rad deploy` parameters](https://edge.docs.radapp.io/reference/cli/rad_deploy/).
+- `-o, --outfile`: The path to the output file where the BicepDeployment resource will be written.
+
+#### BicepDeployment Controller
+
+The `radius-controller` will be updated to include a new controller that reconciles BicepDeployment resources. It will have the following control loop:
+
+1. Check if there is an in-progress operation. If so, check its status:
+    1. If the operation is still in progress, then queue another reconcile operation and continue processing.
+	  2. If the operation completed successfully, then update the `status.phrase` and `status.description` fields as succeeded. Set the contents of the `status.resources` to match the set of deployed resources. This operation is complete, so we can continue processing.
+	  3. If the operation failed, then update the `status.phrase` and `status.description` as failed and continue processing.
```

Comment:
What if only one of the resources failed? Not all of them? Are we going to do a rollback?